snuba/query/snql/joins.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+from typing import MutableMapping, NamedTuple, Optional, Sequence, Union
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinCondition,
+    JoinConditionExpression,
+    JoinRelationship,
+)
+from snuba.query.data_source.simple import Entity as QueryEntity
+from snuba.query.parser.exceptions import ParsingException
+
+
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+    sample_rate: Optional[float] = None
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: EntityTuple
+    relationship: str
+    rhs: EntityTuple
+    data: JoinRelationship
+
+
+class Leaf:
+    def __init__(
+        self,
+        entity: EntityKey,
+        entity_data: EntityTuple,
+        relationship: Optional[JoinRelationship] = None,
+    ) -> None:
+        self.entity = entity
+        self.entity_data = entity_data
+        self.relationship = relationship
+        self.child: Optional[Leaf] = None
+        self.join_conditions: Sequence[JoinCondition] = []
+
+    def push_leaf(self, leaf: Leaf) -> None:
+        # This happens here to ensure we use the correct alias in the columns.
+        self.build_join_conditions(leaf)
+        if not self.child:
+            self.child = leaf
+            return
+
+        old_child = self.child
+        self.child = leaf
+        if leaf.child is None:
+            leaf.child = old_child
+            return
+
+        # iterate down list and push old_child to end
+        head = leaf.child
+        while head.child is not None:
+            head = head.child
+
+        head.child = old_child
+
+    def has_child(self, entity: EntityKey) -> bool:
+        if self.entity == entity:
+            return True
+
+        if self.child:
+            return self.child.has_child(entity)
+
+        return False
+
+    def build_join_conditions(self, rhs: Leaf) -> None:
+        if rhs.relationship is None:
+            return
+
+        join_conditions = []
+        for lhs_column, rhs_column in rhs.relationship.columns:
+            join_conditions.append(
+                JoinCondition(
+                    left=JoinConditionExpression(self.entity_data.alias, lhs_column),
+                    right=JoinConditionExpression(rhs.entity_data.alias, rhs_column),
+                )
+            )
+
+        rhs.join_conditions = join_conditions
+
+
+def build_tree(relationships: Sequence[RelationshipTuple]) -> Leaf:
+    roots: MutableMapping[EntityKey, Leaf] = {}
+    leafs: MutableMapping[EntityKey, Leaf] = {}
+
+    def update_leafs(child: Optional[Leaf]) -> None:
+        while child is not None:
+            leafs[child.entity] = child
+            child = child.child
+
+    for rel in relationships:
+        lhs = Leaf(EntityKey(rel.lhs.name), rel.lhs)
+        rhs = Leaf(EntityKey(rel.rhs.name), rel.rhs, rel.data)
```

Comment:
Is the name of the entity validated previously? Otherwise we should raise a ParsingException if the name provided by the user is not a valid one, otherwise this will raise 500 over an invalid query. But if it is validated previously, could we have the enum in the EntityTuple instead of a bare string ?

Reply:
Yes, the main parser checks the join relationship and validates it.